GAN/main.swift

```diff
@@ -0,0 +1,211 @@
+    exit(-1)
+}
+
+/// Reads MNIST images from specified file path.
+func readMNIST(imagesFile: String) -> Tensor<Float> {
+    print("Reading data.")
+    let images = readFile(imagesFile).dropFirst(16).map { Float($0) }
+    let rowCount = images.count / imageSize
+
+    print("Constructing data tensors.")
+    return Tensor(shape: [rowCount, imageHeight * imageWidth], scalars: images) / 255.0 * 2 - 1
+}
+
+// Models
+
+struct Generator: Layer {
+    var dense1 = Dense<Float>(inputSize: latentSize, outputSize: latentSize * 2,
+                              activation: { leakyRelu($0) })
+    var dense2 = Dense<Float>(inputSize: latentSize * 2, outputSize: latentSize * 4,
+                              activation: { leakyRelu($0) })
+    var dense3 = Dense<Float>(inputSize: latentSize * 4, outputSize: latentSize * 8,
+                              activation: { leakyRelu($0) })
+    var dense4 = Dense<Float>(inputSize: latentSize * 8, outputSize: imageSize,
+                              activation: tanh)
+    
+    var batchnorm1 = BatchNorm<Float>(featureCount: latentSize * 2)
+    var batchnorm2 = BatchNorm<Float>(featureCount: latentSize * 4)
+    var batchnorm3 = BatchNorm<Float>(featureCount: latentSize * 8)
+    
+    @differentiable
+    func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
+        let x1 = batchnorm1(dense1(input))
+        let x2 = batchnorm2(dense2(x1))
+        let x3 = batchnorm3(dense3(x2))
+        return dense4(x3)
+    }
+}
+
+struct Discriminator: Layer {
+    var dense1 = Dense<Float>(inputSize: imageSize, outputSize: 256,
+                              activation: { leakyRelu($0) })
+    var dense2 = Dense<Float>(inputSize: 256, outputSize: 64,
+                              activation: { leakyRelu($0) })
+    var dense3 = Dense<Float>(inputSize: 64, outputSize: 16,
+                              activation: { leakyRelu($0) })
+    var dense4 = Dense<Float>(inputSize: 16, outputSize: 1,
+                              activation: identity)
+    
+    @differentiable
+    func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
+        input.sequenced(through: dense1, dense2, dense3, dense4)
+    }
+}
+
+// Loss functions
+
+@differentiable
+func generatorLoss(fakeLogits: Tensor<Float>) -> Tensor<Float> {
+    sigmoidCrossEntropy(logits: fakeLogits,
+                        labels: Tensor(ones: fakeLogits.shape))
+}
+
+@differentiable
+func discriminatorLoss(realLogits: Tensor<Float>, fakeLogits: Tensor<Float>) -> Tensor<Float> {
+    let realLoss = sigmoidCrossEntropy(logits: realLogits,
+                                       labels: Tensor(ones: realLogits.shape))
+    let fakeLoss = sigmoidCrossEntropy(logits: fakeLogits,
+                                       labels: Tensor(zeros: fakeLogits.shape))
+    return realLoss + fakeLoss
+}
+
+/// Returns `size` samples of noise vector.
+func sampleVector(size: Int) -> Tensor<Float> {
+    Tensor<Float>(randomNormal: [size, latentSize])
```

Comment:
Omit redundant type information since it can be inferred from the return type.
```suggestion
    Tensor(randomNormal: [size, latentSize])
```